src/reducers/composers.js

```diff
@@ -85,33 +75,147 @@ export const updateVisDataComposed = (state, action) => {
 export const addDataToMapComposed = (state, action) => {
 
   const {datasets, options, config} = action.payload;
+  let parsedConfig = config;
+  if (config && config.config && config.version) {
+// if passed in saved config
+    parsedConfig = KeplerGlSchema.parseSavedConfig(config);
+  }
+// Update visState store
+  let mergedState = updateVisDataComposed(state, {datasets, options, config: parsedConfig && parsedConfig.visState});
+  // Update mapState store
+  mergedState = {
+    ...mergedState,
+    mapState: stateMapConfigUpdater(mergedState.mapState, {payload: {mapState: parsedConfig && parsedConfig.mapState}})
+  };
+  // Update mapStyle store
+  mergedState = {
+    ...mergedState,
+    mapStyle: styleMapConfigUpdater(mergedState.mapStyle, {payload: {mapStyle: parsedConfig && parsedConfig.mapStyle}})
+  };
+  return mergedState;
+};
+/**
+ * Remove layer data from layer data
+ * @param state
+ * @param action {datasets, options, config}
+ * @returns state
+ */
+export const removeLayerDataComposed = (state, action) => {
+  // extract mapData
+  const {payload: mapData} = action;
+  const {key: datasetKey} = mapData;
+  const {datasets, layers} = state.visState;
+  // check datasetKey matches with current state 
+  if (!datasets[datasetKey]) {
+    return state;
+  }
+
+// get rows from
+  const {rows} = mapData.datasets.data;
+  // check if rows is present
+  if (!rows) {
+    return state;
+  }
+
+  /* eslint-disable no-unused-vars */
+  const {
+    datasets: {[datasetKey]: dataset, ...newDatasets}
+  } = state.visState;
+  /* eslint-enable no-unused-vars */
+
+  // get layer index 
+  const indexes = layers.reduce((listOfIndexes, layer, index) => {
```

Comment:
If the requirement is only to find the indexes, can you verify the performance using a for loop against reduce, and use the better one
https://codeburst.io/reduce-vs-for-loop-3c1a84e63872

Reply:
`reduce` has been optimised in JS engines to be faster than a `for` loop for arrays of length ~= 15 or more. That being said, what is happening here is strictly a `filter`, plus the performance benefit from using the "better" implementation is negligible and is a micro-optimisation.